src/librustc/ty/walk.rs

```diff
@@ -87,11 +87,14 @@ fn push_subtypes<'tcx>(stack: &mut Vec<Ty<'tcx>>, parent_ty: Ty<'tcx>) {
         ty::TyProjection(ref data) => {
             stack.extend(data.trait_ref.substs.types().rev());
         }
-        ty::TyTrait(ref obj) => {
-            stack.extend(obj.principal.input_types().rev());
-            stack.extend(obj.projection_bounds.iter().map(|pred| {
+        ty::TyDynamic(ref obj, ..) => {
+            match obj.principal() {
+                Some(ref p) => stack.extend(p.input_types().rev()),
+                None => {}
+            }
+            stack.extend(obj.projection_bounds().map(|pred| {
                 pred.0.ty
-            }).rev());
+            }).collect::<Vec<_>>().into_iter().rev()); // FIXME: Don't allocate a vec for reversal.
```

Comment:
I'm confused what you mean by `flat_map` over an iterator of `ExistentialPredicate`. Are you suggesting that we extend from a `filter_map` of `ExistentialPredicate` for the Projection and Trait cases?

(And switch places, so push projection bounds first and then the principal)?